src/IO/S3/Client.h

```diff
@@ -0,0 +1,309 @@
+
+        void GetSendToken() override;
+
+        bool HasSendToken() override;
+
+        void RequestBookkeeping(const Aws::Client::HttpResponseOutcome& httpResponseOutcome) override;
+        void RequestBookkeeping(const Aws::Client::HttpResponseOutcome& httpResponseOutcome, const Aws::Client::AWSError<Aws::Client::CoreErrors>& lastError) override;
+    private:
+        std::shared_ptr<Aws::Client::RetryStrategy> wrapped_strategy;
+    };
+
+    Model::HeadObjectOutcome HeadObject(const HeadObjectRequest & request) const;
+    Model::ListObjectsV2Outcome ListObjectsV2(const ListObjectsV2Request & request) const;
+    Model::ListObjectsOutcome ListObjects(const ListObjectsRequest & request) const;
+    Model::GetObjectOutcome GetObject(const GetObjectRequest & request) const;
+
+    Model::AbortMultipartUploadOutcome AbortMultipartUpload(const AbortMultipartUploadRequest & request) const;
+    Model::CreateMultipartUploadOutcome CreateMultipartUpload(const CreateMultipartUploadRequest & request) const;
+    Model::CompleteMultipartUploadOutcome CompleteMultipartUpload(const CompleteMultipartUploadRequest & request) const;
+    Model::UploadPartOutcome UploadPart(const UploadPartRequest & request) const;
+    Model::UploadPartCopyOutcome UploadPartCopy(const UploadPartCopyRequest & request) const;
+
+    Model::CopyObjectOutcome CopyObject(const CopyObjectRequest & request) const;
+    Model::PutObjectOutcome PutObject(const PutObjectRequest & request) const;
+    Model::DeleteObjectOutcome DeleteObject(const DeleteObjectRequest & request) const;
+    Model::DeleteObjectsOutcome DeleteObjects(const DeleteObjectsRequest & request) const;
+
+private:
+    template <typename... Args>
+    explicit Client(size_t max_redirects_, Args &&... args)
+        : Aws::S3::S3Client(std::forward<Args>(args)...)
+        , max_redirects(max_redirects_)
+        , log(&Poco::Logger::get("S3Client"))
+    {
+        auto * endpoint_provider = dynamic_cast<Aws::S3::Endpoint::S3DefaultEpProviderBase *>(accessEndpointProvider().get());
+        endpoint_provider->GetBuiltInParameters().GetParameter("Region").GetString(explicit_region);
+        std::string endpoint;
+        endpoint_provider->GetBuiltInParameters().GetParameter("Endpoint").GetString(endpoint);
+        detect_region = explicit_region == Aws::Region::AWS_GLOBAL && endpoint.find(".amazonaws.com") != std::string::npos;
+
+        cache = std::make_shared<ClientCache>();
+        ClientCacheRegistry::instance().registerClient(cache);
+    }
+
+    Client(const Client & other)
+        : Aws::S3::S3Client(other)
+        , explicit_region(other.explicit_region)
+        , detect_region(other.detect_region)
+        , max_redirects(other.max_redirects)
+        , log(&Poco::Logger::get("S3Client"))
+    {
+        cache = std::make_shared<ClientCache>(*other.cache);
+        ClientCacheRegistry::instance().registerClient(cache);
+    }
+
+    /// Make regular functions private
```

Comment:
why?

Reply:
If you use requests from SDK the client can't override the region/endpoint so we make the function accepting `Aws::S3::Model::*Request` as private and the only way to call the functions with our client is with our extended requests.